### ISSUE TITLE
clang inconsistent warning on discarding 'nodiscard' result

### ISSUE THREAD

**Issue by wanghan02** [llvm/llvm-project#59616]
Reproducible code could be found below or on [Godbolt](https://godbolt.org/z/c6rrnzoKz).

1. Regular call: Clang, GCC and VS all give warning on discarding `nodiscard` result
2. Inside `noexcept` operator: only VS gives warning on discarding `nodiscard` result

I'm not sure if it is a Clang (or standard?) policy to ignore warnings in unevaluated context. In my opinion consistent warning message is probably better.

```cpp
[[nodiscard]] int bar(int n) noexcept {
    return n;
}

void foo() {
    if constexpr( noexcept(bar(1), 1) ) {} // warning only with VS inside noexcept operator
    [[maybe_unused]] auto n = (bar(1), 1); // warning with clang, gcc and VS
}
```

**Comment by AaronBallman** [comment]
The standard doesn't really say one way or the other (attribute effects are all left up to QoI anyway), but this was an intentional choice in Clang: https://github.com/llvm/llvm-project/blob/a9871772a8b13c1240a95a84a3327f84bb67dddc/clang/lib/Sema/SemaStmt.cpp#L227

We don't warn on unused results in an unevaluated context because there's no reason to assume the user expected the results to be used in the first place, so this is done to reduce potential false positives.
